### PR TITLE
Implements Prettier formatting

### PR DESCRIPTION
**Pull request by RubenRBS**
Moves the precommit handling to husky,
 and the lint calls to lint-staged, so only the staged bits are ran.

This is a scary changeset. I _think_ I didn't break anything,
 but good luck checking that that's actually the case.

So that @partouf does not hate me, I'd wait to merge this
 until after the UI changes are merged in :)

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -1,148 +1,145 @@
 ---
```
Comment (RubenRBS) [inline .eslintrc.yml:1]: Need to delete all the formatting rules from here

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 2]
```diff
@@ -1,34 +1,41 @@
 ---
```
Comment (RubenRBS) [inline .github/ISSUE_TEMPLATE/bug_report.md:1]: Should we add `*.md` to the ignored files by prettier?

File: .prettierrc.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+trailingComma: all
+printWidth: 120
+singleQuote: true
+arrowParens: avoid
+tabWidth: 4
+overrides:
```
Comment (RubenRBS) [inline .prettierrc.yml:6]: Need to double check that this is not breaking anything for `static`. I'd love that the tool supported overriding configs from the current folder as ESLint does

**Review by RubenRBS — COMMENTED** [review]
🙈

File: package.json [hunk 4]
```diff
@@ -1,144 +1,159 @@
...
+        "codecov": "codecov --disable=gcov",
+        "sentry": "npx -p @sentry/cli sentry-cli",
+        "update-browserslist": "npx browserslist@latest -- --update-db",
+        "prepare": "husky install",
+        "format": "prettier --ignore-unknown",
+        "format-fix": "prettier --write --ignore-unknown"
+    },
+    "lint-staged": {
+        "*.{js,css}": [
+            "npm run ci-lint --",
```
Comment (RubenRBS) [inline package.json:154]: ESLint not automatically fixing the code is intended, as this way we keep the old behaviour, just in case

File: lib/base-compiler.js [hunk 5]
```diff
@@ -1406,32 +1488,34 @@ Please select another pass or change filters.`;
...
+            return result;
+        }
+        if (postProcess.length > 0) {
+            return this.execPostProcess(result, postProcess, outputFilename, maxSize);
+        } else {
+            const contents = await fs.readFile(outputFilename);
+            result.asm = contents.toString();
+            return result;
+        }
+    }
```
Comment (RubenRBS) [inline lib/base-compiler.js:1517]: This is a manual refactor

File: .prettierrc.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+trailingComma: all
+printWidth: 120
```
Comment (RubenRBS) [inline .prettierrc.yml:2]: https://prettier.io/docs/en/options.html#print-width gives good reasons as to why this should be 80, but it's not up to me to deceide.

File: .prettierrc.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+singleQuote: true
+arrowParens: avoid
+tabWidth: 4
```
Comment (RubenRBS) [inline .prettierrc.yml:5]: @apmorton gave some good reasons as to why we might want this to be 2 instead of 4 over at Discord some time ago

File: .prettierrc.yml (hunk 3 above)
Comment (RubenRBS) [reply .prettierrc.yml:7]: The site runs fine locally as far as I can tell

File: .prettierrc.yml (hunk 7 above)
Comment (RubenRBS) [reply .prettierrc.yml:5]: I missremembered. Actually, they argued for tabs instead of spaces and linked to https://dev.to/alexandersandberg/why-we-should-default-to-tabs-instead-of-spaces-for-an-accessible-first-environment-101f

Comment (partouf) [reply .prettierrc.yml:5]: I would prefer to stick to whatever we have now in the current source. I think that's 4, so this is fine.

**Comment by partouf** [comment]
pfff, just saw the complete set of changes...

I'll admit I'm not a fan. I just don't think it's smart to do this. This is no fun to merge for any of our outstanding PR's.

Or maybe there's a way to do this in parts? Like only format the file that someone has changed?

**Comment by RubenRBS** [comment]
We can totally do that last point. Most of this comes from manually running `prettier --write .` once. If noone does this, but we keep the pre-commit stuff, we can only run the formatter for the files that someone gets to edit

**Comment by RubenRBS** [comment]
Most of the changes come from thew fact that it ran on `package-lock.json`. It seems that `git checkout 78c0f67c ./package-lock.json && npm i --package-lock-only` should be enough to regenerate it, but I ran out of time to do it and check that it works

File: lib/base-compiler.js (hunk 5 above)
Comment (RubenRBS) [reply lib/base-compiler.js]: Will have to remember to reapply this is main

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 2 above)
Comment (mattgodbolt) [reply .github/ISSUE_TEMPLATE/bug_report.md]: Yeah probably

File: .prettierrc.yml (hunk 6 above)
Comment (mattgodbolt) [reply .prettierrc.yml:2]: I'm ok with 80 actually...it makes GH side-by-side diff work nicely in all cases that I use it :) But... maybe one thing at a time?

File: .prettierrc.yml (hunk 7 above)
Comment (mattgodbolt) [reply .prettierrc.yml:5]: Yeah for now let's stick. @apmorton 's argument for tabs was compelling but ...

**Comment by mattgodbolt** [comment]
> I'll admit I'm not a fan. I just don't think it's smart to do this. 

I'm on board, though I don't have a good solution for doing it "nicely".

> This is no fun to merge for any of our outstanding PR's.

Agreed 100% this is a tricky thing to do right. 

> 
> Or maybe there's a way to do this in parts? Like only format the file that someone has changed?

This is a good idea.

Also, perhaps having one PR that sets up the _infrastructure_ to format code (without actually reformatting). Then...outstanding PRs can mereg _that change_ in and run `make reformat` or whatever on their PRs. That should - in theory - mean they can then merge in the "all the files got reformatted" change without "too much" effort.

So, TLDR; I'd love to do this, but I agree that we need to do it carefully. If it comes to it though I'll be happy to go in and help merge into other people's PRs. We should try and get things closed or merged out.

**Comment by RubenRBS** [comment]
> Also, perhaps having one PR that sets up the infrastructure to format code (without actually reformatting). Then...outstanding PRs can mereg that change in and run make reformat or whatever on their PRs. That should - in theory - mean they can then merge in the "all the files got reformatted" change without "too much" effort.

I've left this exact changes now

File: package.json [hunk 8]
```diff
@@ -1,144 +1,159 @@
...
+        "test": "mocha --recursive",
+        "check": "npm run lint && npm run test",
+        "dev": "cross-env NODE_ENV=DEV node -r esm app.js",
+        "debugger": "cross-env NODE_ENV=DEV node --inspect -r esm app.js --debug",
+        "debug": "cross-env NODE_ENV=DEV node -r esm app.js --debug",
+        "start": "webpack && cross-env NODE_ENV=LOCAL node -r esm app.js",
+        "codecov": "codecov --disable=gcov",
+        "sentry": "npx -p @sentry/cli sentry-cli",
+        "update-browserslist": "npx browserslist@latest -- --update-db",
+        "prepare": "husky install",
```
Comment (mattgodbolt) [inline package.json:148]: +100 to husky. I started using it on other projects! :)

**Comment by partouf** [comment]
Concerning the pre-commit, I sometimes manually delete the `.git/hooks/pre-commit` so I don't have to wait 5 minutes or more for a commit to go through. Can I still do that in this case?

File: .prettierrc.yml (hunk 3 above)
Comment (apmorton) [reply .prettierrc.yml:7]: prettier does search up recursively from the directory of the file being formatted - although I don't think it supports multiple config files.  I believe it stops after finding any config file.

**Comment by apmorton** [comment]
If we are going to apply this piecemeal we should do so manually using the requirePragma option: https://prettier.io/docs/en/options.html#require-pragma

I don't like the idea of formatting being applied as people edit files - it has bad implications for git history and diff readability when reviewing PRs.

The formatting workflow would look something like:
- find some files not modified in any open PRs
- add `@format` comment to the top
- format/commit
- repeat until all files in project are formatted
- remove the `requirePragma` option and (optionally) remove all the `@format` comments

I think we should also strive not to change indentation for now - in _any_ files.
That means:
  - our js stays 4 spaces
  - our json/yaml/etc stays 2 spaces

Also, there is a big disclaimer about leaving `printWidth: 80` here: https://prettier.io/docs/en/options.html#print-width

File: .idea/codeStyles/Project.xml [hunk 9]
```diff
@@ -1,16 +1,40 @@
...
+      <option name="HTML_QUOTE_STYLE" value="Single" />
+      <option name="HTML_ENFORCE_QUOTES" value="true" />
+    </HTMLCodeStyleSettings>
     <JSCodeStyleSettings version="0">
       <option name="FORCE_SEMICOLON_STYLE" value="true" />
+      <option name="SPACE_BEFORE_FUNCTION_LEFT_PARENTH" value="false" />
       <option name="USE_DOUBLE_QUOTES" value="false" />
       <option name="FORCE_QUOTE_STYlE" value="true" />
       <option name="ENFORCE_TRAILING_COMMA" value="WhenMultiline" />
-      <option name="SPACES_WITHIN_IMPORTS" value="true" />
```
Comment (RubenRBS) [inline .idea/codeStyles/Project.xml:8]: I don't know how to set this to true but keep the other one to false in the Idea UI

**Comment by RubenRBS** [comment]
Thanks all. I've made the appropiate changes following your suggestions :)

**Review by partouf — APPROVED** [review]

File: .husky/pre-commit [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npx lint-staged
```
Comment (mattgodbolt) [inline .husky/pre-commit:4]: yay!

**Review by mattgodbolt — APPROVED** [review]
LGTM!
Re: "files not affected in any open PR, I'm not sure that's totally feasible (given some of our large stale PRs), but we can certainly roll out gradually with minimum hassle for active PRs! I'll happily take the flak for the changes :)

Let's do this!

File: etc/scripts/formatcheck.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+def list_modified_paths_in_pr(pr: dict) -> Set[str]:
+    pr_paths = list_pr_files(pr)
+    filtered = {x["filename"] for x in pr_paths if x["status"] == "modified"}
+    return filtered
+
+
+def list_files_under_vc() -> Set[str]:
+    output = subprocess.check_output(
+        ["git", "ls-tree", "-r", "main", "--name-only"]
```
Comment (RubenRBS) [inline etc/scripts/formatcheck.py:70]: TIL about this one when I wrote it :)

**Comment by mattgodbolt** [comment]
So... are we ready for this? If nobody has any worries about merging as-is, I'd be excited to do so :)

File: .prettierrc.yml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+trailingComma: all
+printWidth: 80
+singleQuote: true
+arrowParens: avoid
```
Comment (apmorton) [inline .prettierrc.yml:4]: I think we should remove this and allow it to be defaulted.

From the prettier docs:

    At first glance, avoiding parentheses may look like a better choice because of less visual noise. However, when 
    Prettier removes parentheses, it becomes harder to add type annotations, extra arguments or default values as well  
    as making other changes. Consistent use of parentheses provides a better developer experience when editing real 
    codebases, which justifies the default value for the option.

I tend to agree with their rational

File: .prettierrc.yml [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+tabWidth: 4
+bracketSpacing: false
```
Comment (apmorton) [inline .prettierrc.yml:6]: Also not a huge fan of this myself - but if everyone else prefers it I will subside.

If nothing else, I think having spaces around brackets in object literals makes sense because we already have/enforce spaces around brackets in import statements, single line function bodies, etc.

**Review by apmorton — COMMENTED** [review]
Has anyone looked at what happens with the eslint license header thing when we add the prettier comment to the top of files?

**Comment by RubenRBS** [comment]
As expected, it breaks. It's easy enough to fix with access to the source code, but I'm not sure how to do it with only the config provided by the eslint plugin.

That it, this works for our use case: https://github.com/Stuk/eslint-plugin-header/blob/main/lib/rules/header.js
```diff
diff --git a/lib/rules/header.js b/lib/rules/header.js
index 3504b6f..a3feb35 100644
--- a/lib/rules/header.js
+++ b/lib/rules/header.js
@@ -27,7 +27,11 @@ function excludeShebangs(comments) {
 // check if they are at the start of the file since that is already checked by
 // hasHeader().
 function getLeadingComments(context, node) {
-    var all = excludeShebangs(context.getSourceCode().getAllComments(node.body.length ? node.body[0] : node));
+    var comments = context.getSourceCode().getAllComments(node.body.length ? node.body[0] : node);
+    if (comments.length > 0 && comments[0].type === "Block" && comments[0].value === "*\n * @format\n ") {
+        comments.shift();
+    }
+    var all = excludeShebangs(comments);
     if (all[0].type.toLowerCase() === "block") {
         return [all[0]];
     }
```

But it's questionable at best hah

**Comment by RubenRBS** [comment]
Another idea discussed in Discord was to disable the header plugin for as long as the pragma was needed, which also sounds ok to me
